python/kserve/kserve/storage/storage.py

```diff
@@ -191,17 +192,24 @@ def _download_s3(uri, temp_dir: str):
             # If 'uri' is set to "s3://test-bucket/a/b/c/model.bin", then
             # the downloader will add to temp dir: model.bin
             # (without any subpaths).
-            target_key = (
-                obj.key.rsplit("/", 1)[-1]
-                if bucket_path == obj.key
-                else obj.key.replace(bucket_path, "", 1).lstrip("/")
-            )
+            if bucket_path == obj.key:
+                target_key = obj.key.rsplit("/", 1)[-1]
+                exact_obj_found = True
+            elif obj.key.rsplit("/", 1)[-1].startswith(bucket_path.rsplit("/", 1)[-1]):
+                target_key = obj.key.replace(bucket_path.rsplit("/", 1)[0], "", 1).lstrip("/")
```

Comment:
@andyi2it Is this correct?  if `obj.key` is `churn-pickle-logs` and `bucket_path` is `churn-pickle`, then after replace it becomes `-logs`

Reply:
I think we need to define the expectation here when bucket is `s3://churn-pickle` and obj is `churn-pickle-logs/model.log`, do we still want to download all the files when there is a prefix matching or we do not match at all.

Reply:
@lizzzcai Any thoughts?

Reply:
I feel like matching on a folder or exact file name rather than `prefix` will make things easier. But I am not sure if it will break any `existing user experience` here.


for bucket with following files:
```
test/churn-pickle
test/churn-pickle-logs
test/churn-pickle-report
```
1. When I download `s3://<bucket>/test/churn-pickle` to `/mnt/models`:
only one file is downloaded
```
/mnt/models/churn-pickle
```

2. when I download `s3://<bucket>/test` to `/mnt/models`:
All files under test will be downloaded.
```
/mnt/models/churn-pickle
/mnt/models/churn-pickle-logs
/mnt/models/churn-pickle-report
```

3. When I download `s3://<bucket>/test/churn` to `/mnt/models`:
nothing will be downloaded
```

```

not sure of any corner case here, and for cloud bucket storage, there is no easy way to know if a path is a file or folder.


@yuzisun @andyi2it any comments?

Reply:
Hi @andyi2it,  for 1st scenario, there is a difference from the behaviour in ModelMesh [here](https://github.com/kserve/kserve/issues/2473#issuecomment-1715517612). In modelmesh, only `/mnt/models/churn-pickle` will be downloaded.